src/Features/Core/Portable/Extensions/ExtensionMessageHandlerService.cs

```diff
@@ -0,0 +1,358 @@
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+
+namespace Microsoft.CodeAnalysis.Extensions;
+
+[ExportWorkspaceService(typeof(IExtensionMessageHandlerService)), Shared]
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+internal sealed class ExtensionMessageHandlerService(IExtensionMessageHandlerFactory customMessageHandlerFactory) : IExtensionMessageHandlerService, IDisposable
+{
+    /// <summary>
+    /// Extensions assembly load contexts and loaded handlers, indexed by handler file path. The handlers are indexed by type name.
+    /// </summary>
+    private readonly Dictionary<string, Extension> _extensions = new();
+
+    /// <summary>
+    /// Handlers of document-related messages, indexed by handler message name.
+    /// </summary>
+    private readonly Dictionary<string, IExtensionMessageHandlerWrapper<Document>> _documentHandlers = new();
+
+    /// <summary>
+    /// Handlers of non-document-related messages, indexed by handler message name.
+    /// </summary>
+    private readonly Dictionary<string, IExtensionMessageHandlerWrapper<Solution>> _workspaceHandlers = new();
+
+    private readonly IExtensionMessageHandlerFactory _customMessageHandlerFactory = customMessageHandlerFactory;
+
+    // Used to protect access to _extensions, _handlers, _documentHandlers and CustomMessageHandlerExtension.Assemblies.
+    private readonly object _lockObject = new();
+
+    public ValueTask<RegisterExtensionResponse> RegisterExtensionAsync(
+        Solution solution,
+        string assemblyFilePath,
+        CancellationToken cancellationToken)
+    {
+        var assemblyFileName = Path.GetFileName(assemblyFilePath);
+        var assemblyFolderPath = Path.GetDirectoryName(assemblyFilePath)
+            ?? throw new InvalidOperationException($"Unable to get the directory name for {assemblyFilePath}.");
+
+        var analyzerAssemblyLoaderProvider = solution.Services.GetRequiredService<IAnalyzerAssemblyLoaderProvider>();
+
+        Extension? extension;
+        lock (_lockObject)
+        {
+            // Check if the assembly is already loaded.
+            if (!_extensions.TryGetValue(assemblyFolderPath, out extension))
+            {
+                var analyzerAssemblyLoader = analyzerAssemblyLoaderProvider.CreateNewShadowCopyLoader();
+
+                // Allow this assembly loader to load any dll in assemblyFolderPath.
+                foreach (var dll in Directory.EnumerateFiles(assemblyFolderPath, "*.dll"))
+                {
+                    try
+                    {
+                        // Check if the file is a valid .NET assembly.
+                        AssemblyName.GetAssemblyName(dll);
+                    }
+                    catch
+                    {
+                        // The file is not a valid .NET assembly, skip it.
```

Comment:
is it possible to make this a more narrow exception.  i'm worried we might end up swallowing more important exceptions that indicate a logic error.

Reply:
Based on the [docs](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.assemblyname.getassemblyname?view=net-9.0&devlangs=csharp&f1url=%3FappId%3DDev17IDEF1%26l%3DEN-US%26k%3Dk(System.Reflection.AssemblyName.GetAssemblyName)%3Bk(SolutionItemsProject)%3Bk(DevLang-csharp)%26rd%3Dtrue), I could restrict this to `ArgumentException` and `BadImageFormatException`. But it seems strange to throw an exception here for a file, when we may never be asked to load such file,